### PR TITLE
CUMULUS-1936: Migrate @cumulus/message to TS

### PR DESCRIPTION
**Pull request by markdboyd**
Should come after #1764

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@cumulus/core",
-  "version": "1.0.0",
+  "version": "1.0.1-0",
```
Comment (Jkovarik) [inline package.json:3]: Is there a reason for not using a -alpha designation here?

Comment (markdboyd) [reply package.json:3]: This was unintentional. I'll revert

File: packages/message/src/Build.ts [hunk 2]
```diff
@@ -9,28 +9,49 @@
...
 const uuidv4 = require('uuid/v4');
 
-const createExecutionName = () => uuidv4();
+/**
+ * Generate an execution name.
+ *
+ * @returns {string}
+ * @private
+ */
+const createExecutionName = (): string => uuidv4();
```
Comment (yjpa7145) [inline packages/message/src/Build.ts]: See 4efd8c798

File: packages/message/src/Build.ts [hunk 3]
```diff
@@ -9,28 +9,49 @@
...
 
 /**
  * Build base message.cumulus_meta for a queued execution.
  *
  * @param {Object} params
  * @param {string} params.queueName - An SQS queue name
```
Comment (yjpa7145) [inline packages/message/src/Build.ts:34]: Should be in brackets if it's optional.

File: packages/message/src/Build.ts [hunk 4]
```diff
@@ -9,28 +9,48 @@
...
-  const cumulusMeta = {
+  stateMachine,
+  asyncOperationId,
+  parentExecutionArn
+}: {
+  queueName?: string
+  stateMachine: string,
+  asyncOperationId?: string,
+  parentExecutionArn?: string
+}): Message.CumulusMeta => {
```
Comment (yjpa7145) [inline packages/message/src/Build.ts:52]:
```suggestion
}) => {
```

File: packages/message/src/Build.ts [hunk 5]
```diff
@@ -44,18 +64,23 @@ const buildCumulusMeta = ({
...
-  provider,
-  workflowName
-}) => {
-  const meta = {
+  provider
+}: {
+  workflowName: string
+  collection?: object
+  provider?: object
+}): QueueMessageMeta => {
```
Comment (yjpa7145) [inline packages/message/src/Build.ts:82]:
```suggestion
}) => {
```

File: packages/message/src/Build.ts [hunk 6]
```diff
@@ -44,18 +64,23 @@ const buildCumulusMeta = ({
...
+  const meta:QueueMessageMeta = {
```
Comment (yjpa7145) [inline packages/message/src/Build.ts:83]:
```suggestion
  const meta: QueueMessageMeta = {
```

File: packages/message/src/Build.ts [hunk 7]
```diff
@@ -96,15 +122,26 @@ function buildQueueMessageFromTemplate({
...
+  parentExecutionArn: string,
+  messageTemplate: MessageTemplate,
+  payload: object
+  workflow: Workflow,
+  queueName?: string,
+  asyncOperationId?: string,
+  customCumulusMeta?: object
+  customMeta?: object
+}): Message.CumulusMessage => {
+  const cumulusMeta: Message.CumulusMeta = buildCumulusMeta({
```
Comment (yjpa7145) [inline packages/message/src/Build.ts:137]:
```suggestion
  const cumulusMeta = buildCumulusMeta({
```

File: packages/message/src/Build.ts [hunk 8]
```diff
@@ -96,15 +122,26 @@ function buildQueueMessageFromTemplate({
...
     asyncOperationId,
     parentExecutionArn,
     queueName,
     stateMachine: workflow.arn
   });
 
-  const meta = buildMeta({
+  const meta: QueueMessageMeta = buildMeta({
```
Comment (yjpa7145) [inline packages/message/src/Build.ts:144]:
```suggestion
  const meta = buildMeta({
```

File: packages/message/src/Collections.ts [hunk 9]
```diff
@@ -21,29 +22,26 @@ const get = require('lodash/get');
  *
  * @alias module:Collections
  */
-function constructCollectionId(name, version) {
-  return `${name}___${version}`;
-}
+export const constructCollectionId = (name:string, version:string): string =>
```
Comment (yjpa7145) [inline packages/message/src/Collections.ts:25]:
```suggestion
export const constructCollectionId = (name: string, version: string): string =>
```

File: packages/message/src/Collections.ts [hunk 10]
```diff
@@ -21,29 +22,26 @@ const get = require('lodash/get');
...
+  `${name}___${version}`;
 
 /**
  * Get collection ID from execution message.
  *
- * @param {Object} message       - An execution message
+ * @param {Message.CumulusMessage} message - An execution message
```
Comment (yjpa7145) [inline packages/message/src/Collections.ts:31]:
```suggestion
 * @param {CumulusMessage} message - An execution message
```

File: packages/message/src/Executions.ts [hunk 11]
```diff
@@ -22,7 +23,10 @@ const isString = require('lodash/isString');
  *
  * @alias module:Executions
  */
-const buildExecutionArn = (stateMachineArn, executionName) => {
+export const buildExecutionArn = (
+  stateMachineArn: string,
+  executionName: string
+): string | null => {
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:29]:
```suggestion
) => {
```

File: packages/message/src/Executions.ts [hunk 12]
```diff
@@ -38,11 +42,11 @@ const buildExecutionArn = (stateMachineArn, executionName) => {
  *
  * @alias module:Executions
  */
-function getExecutionUrlFromArn(executionArn) {
+export const getExecutionUrlFromArn = (executionArn: string): string => {
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:45]:
```suggestion
export const getExecutionUrlFromArn = (executionArn: string) => {
```

File: packages/message/src/Executions.ts [hunk 13]
```diff
@@ -52,7 +56,9 @@ function getExecutionUrlFromArn(executionArn) {
  *
  * @alias module:Executions
  */
-const getStateMachineArnFromExecutionArn = (executionArn) => {
+export const getStateMachineArnFromExecutionArn = (
+  executionArn: string
+): string | null => {
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:61]:
```suggestion
) => {
```

File: packages/message/src/Executions.ts [hunk 14]
```diff
@@ -62,12 +68,15 @@ const getStateMachineArnFromExecutionArn = (executionArn) => {
...
+ * @param {Message.CumulusMessage} message - A workflow message object
  * @returns {string} An execution name
+ * @throws {Error} if there is no execution name
  *
  * @alias module:Executions
  */
-const getMessageExecutionName = (message) => {
+export const getMessageExecutionName = (
+  message: Message.CumulusMessage
+): string => {
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:79]:
```suggestion
) => {
```

File: packages/message/src/Executions.ts [hunk 15]
```diff
@@ -62,12 +68,15 @@ const getStateMachineArnFromExecutionArn = (executionArn) => {
 /**
  * Get the execution name from a workflow message.
  *
- * @param {Object} message - A workflow message object
+ * @param {Message.CumulusMessage} message - A workflow message object
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:71]:
```suggestion
 * @param {CumulusMessage} message - A workflow message object
```

File: packages/message/src/Executions.ts [hunk 16]
```diff
@@ -78,12 +87,15 @@ const getMessageExecutionName = (message) => {
 /**
  * Get the state machine ARN from a workflow message.
  *
- * @param {Object} message - A workflow message object
+ * @param {Message.CumulusMessage} message - A workflow message object
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:90]:
```suggestion
 * @param {CumulusMessage} message - A workflow message object
```

File: packages/message/src/Executions.ts [hunk 17]
```diff
@@ -78,12 +87,15 @@ const getMessageExecutionName = (message) => {
...
  * @returns {string} A state machine ARN
+ * @throws {Error} if there is not state machine ARN
  *
  * @alias module:Executions
  */
-const getMessageStateMachineArn = (message) => {
+export const getMessageStateMachineArn = (
+  message: Message.CumulusMessage
+): string => {
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:98]:
```suggestion
) => {
```

File: packages/message/src/Executions.ts [hunk 18]
```diff
@@ -94,12 +106,14 @@ const getMessageStateMachineArn = (message) => {
...
- * @param {Object} message - A workflow message object
+ * @param {Message.CumulusMessage} message - A workflow message object
  * @returns {null|string} A state machine execution ARN
  *
  * @alias module:Executions
  */
-const getMessageExecutionArn = (message) => {
+export const getMessageExecutionArn = (
+  message: Message.CumulusMessage
+): null | string => {
```
Comment (yjpa7145) [inline packages/message/src/Executions.ts:116]:
```suggestion
) => {
```

File: packages/message/src/Granules.ts [hunk 19]
```diff
@@ -8,19 +8,18 @@
...
  * @alias module:Granules
  */
-const getMessageGranules = (message) => get(message, 'payload.granules');
-
-module.exports = {
-  getMessageGranules
-};
+export const getMessageGranules = (
+  message: Message.CumulusMessage
+): Array<object> | undefined => get(message, 'payload.granules');
```
Comment (yjpa7145) [inline packages/message/src/Granules.ts:25]:
```suggestion
): unknown[] | undefined => get(message, 'payload.granules');
```

File: packages/message/src/Granules.ts [hunk 20]
```diff
@@ -8,19 +8,18 @@
...
 
-const get = require('lodash/get');
+import get from 'lodash/get';
+import { Message } from '@cumulus/types';
 
 /**
  * Get granules from execution message.
  *
- * @param {Object} message - An execution message
+ * @param {Message.CumulusMessage} message - An execution message
```
Comment (yjpa7145) [inline packages/message/src/Granules.ts:17]:
```suggestion
 * @param {CumulusMessage} message - An execution message
```

File: packages/types/package.json [hunk 21]
```diff
@@ -11,6 +11,9 @@
   "engines": {
     "node": ">=12.18.0"
   },
+  "publishConfig": {
```
Comment (yjpa7145) [inline packages/types/package.json:14]: So is `publishConfig` actually needed?

File: packages/message/nyc.config.js [hunk 22]
```diff
@@ -2,5 +2,6 @@
 
 module.exports = {
   extends: '../../nyc.config.js',
-  include: ['*.js']
+  include: ['*.js'],
+  'exclude-after-remap': false
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply packages/message/nyc.config.js:6]: This is probably just a bad copy/paste

File: packages/message/src/Collections.ts (hunk 10 above)
Comment (markdboyd) [reply packages/message/src/Collections.ts:31]: The actual type is `Message.CumulusMessage` though?

File: packages/types/package.json (hunk 21 above)
Comment (markdboyd) [reply packages/types/package.json:14]: Yeah, NPM gives an error about not being able to publish "private" packages without it. Found out when publishing `@cumulus/types`

File: packages/message/src/Collections.ts (hunk 10 above)
Comment (yjpa7145) [reply packages/message/src/Collections.ts:31]: The type is `CumulusMessage`, ou just happen to be exporting it under `Message`. As it's currently defined, there is no namespace associated with that type.

File: packages/message/src/Collections.ts [hunk 23]
```diff
@@ -21,29 +22,26 @@ const get = require('lodash/get');
  *
  * @alias module:Collections
  */
-function constructCollectionId(name, version) {
-  return `${name}___${version}`;
-}
+export const constructCollectionId = (name: string, version: string): string =>
```
Comment (yjpa7145) [inline packages/message/src/Collections.ts:25]:
```suggestion
export const constructCollectionId = (name: string, version: string) =>
```

File: packages/message/src/Collections.ts [hunk 24]
```diff
@@ -21,29 +22,26 @@ const get = require('lodash/get');
...
  * @returns {string | undefined} - A collection ID or undefined if
  *                                 message.meta.collection isn't
  *                                 present
  *
  * @alias module:Collections
  */
-const getCollectionIdFromMessage = (message) => {
+export const getCollectionIdFromMessage = (
+  message: Message.CumulusMessage
+): string | undefined => {
```
Comment (yjpa7145) [inline packages/message/src/Collections.ts:40]:
```suggestion
) => {
```

File: packages/message/src/Queue.ts [hunk 25]
```diff
@@ -8,33 +8,38 @@
...
- * @returns {string} An SQS queue name
+ * @returns {string|undefined} An SQS queue name or undefined
  *
  * @alias module:Queue
  */
-const getQueueNameByUrl = (message, queueUrl) => {
+export const getQueueNameByUrl = (
+  message: Message.CumulusMessage,
+  queueUrl: string
+): string | undefined => {
```
Comment (yjpa7145) [inline packages/message/src/Queue.ts:28]:
```suggestion
) => {
```

File: packages/message/src/Queue.ts [hunk 26]
```diff
@@ -8,33 +8,38 @@
...
  *
- * @param {Object} message - A workflow message object
+ * @param {Message.CumulusMessage} message - A workflow message object
  * @returns {string} A queue name
+ * @throws {Error} if no queue name in the message
  *
  * @alias module:Queue
  */
-const getQueueName = (message) => {
+export const getQueueName = (message: Message.CumulusMessage): string => {
```
Comment (yjpa7145) [inline packages/message/src/Queue.ts:42]:
```suggestion
export const getQueueName = (message: Message.CumulusMessage) => {
```

File: packages/message/src/Queue.ts [hunk 27]
```diff
@@ -45,13 +50,17 @@ const getQueueName = (message) => {
...
  * @returns {number} Count of the maximum executions for the queue
+ * @throws {Error} if no maximum executions can be found
  *
  * @alias module:Queue
  */
-const getMaximumExecutions = (message, queueName) => {
+export const getMaximumExecutions = (
+  message: Message.CumulusMessage,
+  queueName: string
+): number => {
```
Comment (yjpa7145) [inline packages/message/src/Queue.ts:63]:
```suggestion
) => {
```

File: packages/message/src/Queue.ts [hunk 28]
```diff
@@ -62,12 +71,12 @@ const getMaximumExecutions = (message, queueName) => {
...
  * Determine if there is a queue and queue execution limit in the message.
  *
- * @param {Object} message - A workflow message object
+ * @param {Message.CumulusMessage} message - A workflow message object
  * @returns {boolean} True if there is a queue and execution limit.
  *
  * @alias module:Queue
  */
-const hasQueueAndExecutionLimit = (message) => {
+export const hasQueueAndExecutionLimit = (message: Message.CumulusMessage): boolean => {
```
Comment (yjpa7145) [inline packages/message/src/Queue.ts:79]:
```suggestion
export const hasQueueAndExecutionLimit = (message: Message.CumulusMessage) => {
```

File: packages/message/src/StepFunctions.ts [hunk 29]
```diff
@@ -58,30 +63,39 @@ const pullStepFunctionEvent = async (event) => {
...
  * @param {string} stepName - Name of the step
  * @returns {Promise<Object>} Parsed and updated event with target path replaced by remote message
  *
  * @async
  * @alias module:StepFunctions
  */
-const parseStepMessage = async (stepMessage, stepName) => {
-  let parsedStepMessage = stepMessage;
+
+type CMAMessage = {
```
Comment (yjpa7145) [inline packages/message/src/StepFunctions.ts:74]: This should be exported, and should probably live somewhere else.

Comment (markdboyd) [reply packages/message/src/StepFunctions.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

File: packages/message/src/Queue.ts (hunk 28 above)
Comment (markdboyd) [reply packages/message/src/Queue.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

File: packages/message/src/Queue.ts (hunk 27 above)
Comment (markdboyd) [reply packages/message/src/Queue.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

File: packages/message/src/Queue.ts (hunk 26 above)
Comment (markdboyd) [reply packages/message/src/Queue.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

File: packages/message/src/Queue.ts (hunk 25 above)
Comment (markdboyd) [reply packages/message/src/Queue.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

File: packages/message/src/Collections.ts (hunk 24 above)
Comment (markdboyd) [reply packages/message/src/Collections.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

File: packages/message/src/Collections.ts (hunk 23 above)
Comment (markdboyd) [reply packages/message/src/Collections.ts]: ee0fc38ba7f6b29a7a39153e0e60e41406d6a5c4

**Pull request merged**